src/tests/Microsoft.PowerFx.Core.Tests/AnalyzeThreadSafety.cs

```diff
@@ -0,0 +1,177 @@
+            // `int Prop {get; } = 123`      // readonly backing field! 
+            var errors = new List<string>();
+
+            var total = 0;
+            foreach (var type in asm.GetTypes())
+            {
+                // Reflecting over fields will also find compiler-generated "backing fields" from static properties. 
+                foreach (var field in type.GetFields(BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic))
+                {
+                    var name = $"{type.Name}.{field.Name}";
+                    total++;
+
+                    if (field.Attributes.HasFlag(FieldAttributes.Literal))
+                    {
+                        continue;  // 'const' keyword, 
+                    }
+
+                    if (type.Name.Contains("<>") ||
+                        type.Name.Contains("<PrivateImplementationDetails>"))
+                    {
+                        continue; // exclude compiler generated closures. 
+                    }
+
+                    // Field has is protected by a lock.
+                    if (field.GetCustomAttributes<ThreadSafeProtectedByLockAttribute>().Any())
+                    {
+                        continue;
+                    }
+
+                    if (bugsFieldType != null && bugsFieldType.Contains(field.FieldType))
+                    {
+                        continue;
+                    }
+
+                    if (bugNames != null && bugNames.Contains(name))
+                    {
+                        continue;
+                    }
+
+                    if (field.GetCustomAttribute<ThreadStaticAttribute>() != null)
+                    {
+                        // If field is marked [ThreadStatic], then each thread gets its own copy.
+                        // It also implies the author thought about threading. 
+                        continue;
+                    }
+
+                    if (IsFieldVolatile(field))
+                    {
+                        // If a field was marked volatile, then assume the author thought through the threading.
+                        continue;
+                    }
+
+                    // Is it readonly? Const?
+                    if (!field.Attributes.HasFlag(FieldAttributes.InitOnly))
+                    {
+                        // Mutable static field! That's bad.  
+                        errors.Add($"{name} is not readonly");
+                        continue;
+                    }
+
+                    // Is it a
+                    if (!IsTypeImmutable(field.FieldType))
+                    {
+                        errors.Add($"{name} readonly, but still a mutable type {field.FieldType}");
+                    }
+
+                    // Safe! The static field is readonly and set to an immutable object. 
+                }
+            }
+
+            // Sanity check that we actually ran the test. 
+            Assert.True(total > 10, "failed to find fields");
```

Comment:
>10 [](http://example.com/codeflow?start=32&length=2)

What is this 10? What does this represent?

Reply:
I just wanted to make sure we're actually scanning something and the test is doing meaningful work. 
I didn't want a bug where we failed to scan fields (eg, wrong reflection flags came up empty) and then passed because we didn't do any work.